src/shims/tls.rs

```diff
@@ -292,22 +324,37 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
 
     /// Schedule an active thread's TLS destructor to run on the active thread.
     /// Note that this function does not run the destructors itself, it just
-    /// schedules them one by one each time it is called.
+    /// schedules them one by one each time it is called and reenables the
+    /// thread so that it can be executed normally by the main execution loop.
     ///
     /// FIXME: we do not support yet deallocation of thread local statics.
-    fn schedule_tls_dtors_for_active_thread(&mut self) -> InterpResult<'tcx> {
+    /// Issue: https://github.com/rust-lang/miri/issues/1369
+    fn schedule_next_tls_dtor_for_active_thread(&mut self) -> InterpResult<'tcx> {
         let this = self.eval_context_mut();
         let active_thread = this.get_active_thread()?;
 
-        if this.tcx.sess.target.target.target_os == "windows" {
-            if !this.machine.tls.dtors_running.contains(&active_thread) {
-                this.machine.tls.dtors_running.insert(active_thread);
+        let scheduled_next = if this.tcx.sess.target.target.target_os == "windows" {
+            if !this.machine.tls.set_dtors_running_for_thread(active_thread) {
                 this.schedule_windows_tls_dtors()?;
+                true
+            } else {
+                false
             }
         } else {
-            this.machine.tls.dtors_running.insert(active_thread);
-            this.schedule_macos_global_tls_dtors()?;
-            this.schedule_pthread_tls_dtors()?;
+            this.machine.tls.set_dtors_running_for_thread(active_thread);
+            // The macOS thread wide destructor runs "before any TLS slots get
+            // freed", so do that first.
+            if this.schedule_macos_tls_dtor()? {
+                true
+            } else {
+                this.schedule_pthread_tls_dtors()?
```

Comment:
Should probably be renamed to `schedule_pthread_tls_dtor` or `schedule_next_pthread_tls_dtor` or so.